### PR TITLE
Redesign manager

### PR DESCRIPTION
**Pull request by romkatsu**
| Q             | A
| ------------- | ---
| Is bugfix?    | ❌
| New feature?  | ❌
| Breaks BC?    | ✔️
| Tests pass?   | ✔️
| Fixed issues  | #35 

- PhpManager is divided into storage and Manager
- Example of implementing [another storage(draft)](https://github.com/romkatsu/rbac-storage-ar)
- redesigned Manager
- removed custom exceptions

### CODE REVIEW

**Comment by romkatsu** [comment]
[Redesign RBAC](https://trello.com/c/V4htEpd4/64-redesign-rbac)

File: src/Manager.php [hunk 1]
```diff
@@ -0,0 +1,704 @@
...
+ * Manager represents an authorization manager that stores authorization
+ * For more details and usage information on Manager, see the [guide article on security authorization](guide:security-authorization).
+ */
+final class Manager implements AccessCheckerInterface
+{
+    private Storage $storage;
+    private RuleFactoryInterface $ruleFactory;
+
+    /**
+     * @var array a list of role names that are assigned to every user automatically without calling [[assign()]].
```
Comment (rustamwin) [inline src/Manager.php:21]:
```suggestion
     * @var array a list of role names that are assigned to every user automatically without calling {@see assign()}.
```

Comment (romkatsu) [reply src/Manager.php]: fixed

File: src/Item.php [hunk 2]
```diff
@@ -44,11 +44,18 @@ public function __construct(string $name)
 
     abstract public function getType(): string;
 
+    abstract public function canBeParentOfItem(Item $child): bool;
```
Comment (samdark) [inline src/Item.php:47]:
```suggestion
    abstract public function canBeParentOf(Item $child): bool;
```

File: src/Item.php [hunk 3]
```diff
@@ -44,11 +44,18 @@ public function __construct(string $name)
...
+
     public function getName(): string
     {
         return $this->name;
     }
 
+    public function isEqualName(string $name): bool
```
Comment (samdark) [inline src/Item.php:54]: I don't think that deserves a method. `->getName() === $name` should be enough.

File: src/Item.php [hunk 4]
```diff
@@ -80,6 +87,11 @@ public function getRuleName(): ?string
         return $this->ruleName;
     }
 
+    public function hasRuleName(): bool
```
Comment (samdark) [inline src/Item.php:90]: I don't think that deserves a method. `->getRuleName() !== null` should be enough.

File: src/Manager.php [hunk 5]
```diff
@@ -0,0 +1,703 @@
...
+     *
+     * @param Item $parent
+     * @param Item $child
+     *
+     * @throws RuntimeException
+     * @throws InvalidArgumentException
+     */
+    public function addChild(Item $parent, Item $child): void
+    {
+        if (!$this->hasItem($parent->getName()) || !$this->hasItem($child->getName())) {
```
Comment (samdark) [inline src/Manager.php:78]: How about separating these two exceptions? Would make errors more useful.

File: src/Manager.php [hunk 6]
```diff
@@ -0,0 +1,703 @@
...
+            );
+        }
+
+        if ($parent->isEqualName($child->getName())) {
+            throw new InvalidArgumentException(sprintf('Cannot add "%s" as a child of itself.', $parent->getName()));
+        }
+
+        if (!$parent->canBeParentOfItem($child)) {
+            throw new InvalidArgumentException(
+                sprintf('Can not add "%s" role as a child of "%s" permission.', $child->getName(), $parent->getName())
```
Comment (samdark) [inline src/Manager.php:90]: Here we are assuming that `canBeParentOfItem` may fail for a single reason. If so, it worth moving the check here and deleting `Item::canBeParentOfItem()`.

File: src/PhpStorage.php [hunk 7]
```diff
@@ -0,0 +1,657 @@
...
+        $this->itemFile = $directory . DIRECTORY_SEPARATOR . $itemFile;
+        $this->assignmentFile = $directory . DIRECTORY_SEPARATOR . $assignmentFile;
+        $this->ruleFile = $directory . DIRECTORY_SEPARATOR . $ruleFile;
+        $this->load();
+    }
+
+    /**
+     * @return Item[]
+     */
+    public function getItems(): array
```
Comment (samdark) [inline src/PhpStorage.php:92]: All method phpdoc from here should be moved up to interface. There should be no need to have it in this class.

File: src/PhpStorage.php [hunk 8]
```diff
@@ -0,0 +1,657 @@
...
+
+        $this->addItem($item);
+    }
+
+    private function updateItemName(string $name, Item $item): void
+    {
+        if ($this->hasItem($item->getName())) {
+            throw new InvalidArgumentException(
+                sprintf(
+                    'Unable to change the item name. The name "%s" is already used by another item.',
```
Comment (samdark) [inline src/PhpStorage.php:324]: Can this check be moved to manager?

File: src/Item.php (hunk 2 above)
Comment (romkatsu) [reply src/Item.php]: fixed

File: src/Item.php (hunk 3 above)
Comment (romkatsu) [reply src/Item.php]: fixed

File: src/Item.php (hunk 4 above)
Comment (romkatsu) [reply src/Item.php]: fixed

File: src/Manager.php (hunk 5 above)
Comment (romkatsu) [reply src/Manager.php]: fixed

File: src/Manager.php (hunk 6 above)
Comment (romkatsu) [reply src/Manager.php:96]: fixed

File: src/PhpStorage.php (hunk 7 above)
Comment (romkatsu) [reply src/PhpStorage.php:89]: fixed

File: src/PhpStorage.php (hunk 8 above)
Comment (romkatsu) [reply src/PhpStorage.php]: moved

File: src/Manager.php [hunk 9]
```diff
@@ -0,0 +1,730 @@
...
+     * Returns a value indicating whether the child already exists for the parent.
+     *
+     * @param Item $parent
+     * @param Item $child
+     *
+     * @return bool whether `$child` is already a child of `$parent`
+     */
+    public function hasChild(Item $parent, Item $child): bool
+    {
+        return isset($this->storage->getChildrenByName($parent->getName())[$child->getName()]);
```
Comment (rustamwin) [inline src/Manager.php:160]:
```suggestion
        return array_key_exists($child->getName(), $this->storage->getChildrenByName($parent->getName()));
```

File: src/Storage.php [hunk 10]
```diff
@@ -0,0 +1,163 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Yiisoft\Rbac;
+
+interface Storage
```
Comment (rustamwin) [inline src/Storage.php:7]:
```suggestion
interface StorageInterface
```

File: src/Manager.php (hunk 9 above)
Comment (romkatsu) [reply src/Manager.php]: fixed

File: src/Storage.php (hunk 10 above)
Comment (romkatsu) [reply src/Storage.php]: rename

File: src/StorageInterface.php [hunk 11]
```diff
@@ -0,0 +1,214 @@
...
+    /**
+     * Removes a role, permission or rule from the RBAC system.
+     * @param Item $item Item to remove.
+     */
+    public function removeItem(Item $item): void;
+
+    /**
+     * @return array
+     */
+    public function getChildren(): array;
```
Comment (samdark) [inline src/StorageInterface.php:54]: That doesn't looks efficient for storages such as SQL.

File: src/StorageInterface.php [hunk 12]
```diff
@@ -0,0 +1,214 @@
...
+     * @param string $userId The user ID.
+     * @param Item $item Role to assign.
+     */
+    public function addAssignment(string $userId, Item $item): void;
+
+    /**
+     * @param string $name
+     * @return bool
+     */
+    public function assignmentExist(string $name): bool;
```
Comment (samdark) [inline src/StorageInterface.php:164]: What is a name here? Name of the item that is being assigned?

File: src/StorageInterface.php [hunk 13]
```diff
@@ -0,0 +1,215 @@
...
+     * @param Item $item Role to assign.
+     */
+    public function addAssignment(string $userId, Item $item): void;
+
+    /**
+     * Returns whether there is assignment for a named role or permission.
+     * @param string $name Name of the role or the permission.
+     * @return bool Whether there is assignment.
+     */
+    public function assignmentExist(string $name): bool;
```
Comment (samdark) [inline src/StorageInterface.php:165]: Does not seem to be used anywhere except tests. It it usable as public method?

File: src/Manager.php [hunk 14]
```diff
@@ -0,0 +1,715 @@
...
+        if ($rule->getName() !== $name) {
+            $this->storage->removeRule($name);
+        }
+        $this->storage->addRule($rule);
+    }
+
+    /**
+     * Set default roles.
+     *
+     * @param string[]|callable $roles either array of roles or a callable returning it
```
Comment (samdark) [inline src/Manager.php:403]: Seems it can be only a closure, not array-callable.

**Comment by samdark** [comment]
Great improvement overall!

**Pull request merged**